File: app/src/main/java/app/recall/capture/MailThread.kt
package app.recall.capture

/**
 * An email conversation is one thread (sender + subject), not everything from a sender:
 * ten different alerts or PRs from one address are ten separate things.
 */
object MailThread {
    private val PREFIX = Regex("""^\s*((re|fwd?|aw|wg)\s*(\[\d+])?\s*:\s*)+""", RegexOption.IGNORE_CASE)
    private val SPACE = Regex("""\s+""")

    /** "Re: Fwd: [Cosmos] ETL run failed…" -> "[cosmos] etl run failed" */
    fun subject(text: String): String =
        text.lineSequence().firstOrNull { it.isNotBlank() }.orEmpty()
            .replace(PREFIX, "").replace(SPACE, " ").trim().trimEnd('…', '.').trim()
            .lowercase().take(60).trim()

    fun key(pkg: String, sender: String?, text: String): String = "$pkg|${sender.orEmpty()}|${subject(text)}"
}

File: ios/RecallCore/Sources/RecallCore/Understand/MailThread.swift
import Foundation

/**
 * An email conversation is one thread (sender + subject), not everything from a sender:
 * ten different alerts or PRs from one address are ten separate things.
 */
public enum MailThread {
    private static let PREFIX = RulePattern(#"^\s*((re|fwd?|aw|wg)\s*(\[\d+])?\s*:\s*)+"#, ignoreCase: true)
    private static let SPACE = RulePattern(#"\s+"#)

    /** "Re: Fwd: [Beacon] ETL run failed…" -> "[beacon] etl run failed" */
    public static func subject(_ text: String) -> String {
        let first = text.ktLines().first { !$0.ktIsBlank } ?? ""
        let s = SPACE.replace(PREFIX.replace(first, with: ""), with: " ").ktTrim().ktTrimEnd("…", ".").ktTrim()
        return s.lowercased().ktTake(60).ktTrim()
    }

    public static func key(_ pkg: String, _ sender: String?, _ text: String) -> String {
        "\(pkg)|\(sender ?? "")|\(subject(text))"
    }
}
